assets/js/workflow-reducer.js

```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
-  var state = store.getState();
+  let newSelectedId = null;
+  let state = store.getState();
   if (id_to_delete === state.selected_wf_module) {
 
     // Find id of previous in stack
-    var wf_modules = state.workflow.wf_modules;
-    var new_selected_id = null;
-    for (var wfm of wf_modules) {
+    let wf_modules = state.workflow.wf_modules;
+    for (let wfm of wf_modules) {
       if (wfm.id === id_to_delete)
         break;
-      new_selected_id = wfm.id;
+      newSelectedId = wfm.id;
     }
 
     // if we are deleting first module, set to new first module if any
-    if (new_selected_id === null) {
-      if (wf_modules.length > 1)
-        new_selected_id = wf_modules[1].id;
-      else
-        new_selected_id = null;
+    if (newSelectedId === null && wf_modules.length > 1) {
+      newSelectedId = wf_modules[1].id;
     }
-
-    store.dispatch(changeSelectedWfModuleAction(new_selected_id))
   }
 
-  return (
-    api.deleteModule(id_to_delete)
-    .then( reloadWorkflowAction )
-  )
-}
-
-export function wfModuleStatusAction(wfModuleID, status, error_msg='') {
-  return {
-    type : MODULE_STATUS_CHANGE,
-    id : wfModuleID,
-    status : status,
-    error_msg: error_msg
+  // If the selected module isn't changing, set it to the current selected module.
+  if (newSelectedId === null) {
+    newSelectedId = state.selected_wf_module;
   }
-}
 
-export function changeSelectedWfModuleAction(wfModuleID) {
-  return {
-    type : SELECTED_MODULE_CHANGE,
-    id : wfModuleID,
+  return function (dispatch) {
+    return (
+      // Set the new selected module before deleting to avoid errors.
+      // We do this even if the selected module isn't changing to avoid
+      // writing a subtly tricky conditional here.
+      dispatch(setSelectedWfModuleAction(newSelectedId)).then(() => {
+      // Remove the module
+      dispatch({
+        type: DELETE_MODULE,
+        payload: {
+          promise: api.deleteModule(id_to_delete),
+          data: {
+            wf_module_id: id_to_delete
+          }
+        }
+      });
+    }));
   }
 }
+registerReducerFunc(DELETE_MODULE + '_PENDING', (state, action) => {
+  let wfModuleIdx = findIdxByProp(
+    state.workflow.wf_modules,
+    'id',
+    action.payload.wf_module_id
+  );
+
+  return update(state, {
+    workflow: {
+      wf_modules: {$splice: [[wfModuleIdx, 1]] }
+    }
+  });
+});
+
 
-export function toggleModuleCollapsedAction(wfModuleID, isCollapsed) {
+// SET_SELECTED_MODULE
+// Set the selected module in the workflow
+export function setSelectedWfModuleAction(wfModuleID) {
+  let workflowID = store.getState().workflow.id;
   return {
-    type : TOGGLE_MODULE_COLLAPSED,
-    id : wfModuleID,
-    isCollapsed
+    type : SET_SELECTED_MODULE,
+    payload : {
+      promise: api.setSelectedWfModule(workflowID, wfModuleID).then((response) => {
+        return {
+          wf_module_id: response.selected_wf_module
```

Comment:
Why is this needed? I would think this value only appears in _FULFILLED, and we don't bother handling it.